### PR TITLE
Update ResNet example to realistic image

### LINKED ISSUE #13: Update resnet benchmarking
Once #11 is merged, it would be great to use the new timings with the [more realistic resnet example](https://github.com/Cambridge-ICCS/fortran-pytorch-lib/pull/38).

### PR DESCRIPTION
**Pull request by ElliottKasoar**
Resolves #13

To do:
- [x] Remove hardcoded datafile (do we want to save the binary in the repo, similar to the example, or have it generated by the forpy section?)
- [x] Update forpy benchmark
- [x] Separate out precision module (only the Fortran types, or also the C types?)

Updates ResNet benchmarking, based on [updated example in FTorch](https://github.com/Cambridge-ICCS/FTorch/tree/main/examples/2_ResNet18). This includes:

- Updated benchmarker_resnet_forpy.f90 and benchmarker_resnet_torch.f90 to read in image_tensor.dat, a binary storing the same dog.jpg tensor as is used in the FTorch example
- precision module to define single and double precision in both benchmarker_resnet_forpy.f90 and benchmarker_resnet_torch.f90
- Updated utils to include `assert_real`, as is defined in the FTorch example
- Updated python scripts, which are essentially copies of those in the FTorch example, with minor changes e.g. keeping `initialize_ts` and `compute` functions, and defining a default torch precision.

### CODE REVIEW

File: benchmarker_utils.f90 [hunk 1]
```diff
@@ -40,6 +40,36 @@ subroutine assert_real_2d(a, b, test_name, rtol_opt)
 
   end subroutine assert_real_2d
 
+  subroutine assert_real(a, b, test_name, rtol_opt)
```
Comment (TomMelt) [inline benchmarker_utils.f90:43]: can we do something like `print_time_stats` and make an interface for the assertions. `assert_real` could be the interface and we could make assert_real_1d for singular values, for example. Then in the main code we can just use `asset_real` for any comparison involving reals.

File: benchmarker_utils.f90 [hunk 2]
```diff
@@ -40,6 +40,36 @@ subroutine assert_real_2d(a, b, test_name, rtol_opt)
...
+    implicit none
+
+    character(len=*) :: test_name
+    real, intent(in) :: a, b
+    real, optional :: rtol_opt
+    real :: relative_error, rtol
+
+    character(len=15) :: pass, fail
+
+    fail = char(27)//'[31m'//'FAILED'//char(27)//'[0m'
```
Comment (TomMelt) [inline benchmarker_utils.f90:54]: Not necessary for this PR but we could think about ways to remove code duplication in the assertion statements.

**Review by TomMelt — CHANGES_REQUESTED** [review]
If we can move assertions into an interface then I am happy to merge.

The second comment (w.r.t reducing code duplication) is just a comment to think about for future.

File: benchmarker_utils.f90 (hunk 1 above)
Comment (ElliottKasoar) [reply benchmarker_utils.f90:47]: Good idea, thanks! Done

File: benchmarker_utils.f90 (hunk 2 above)
Comment (ElliottKasoar) [reply benchmarker_utils.f90:58]: Thanks, yes I agree. I've added it as a to-do for the general tidy up in #24

**Review by TomMelt — APPROVED** [review]
LGTM

**Pull request merged**